### ISSUE TITLE
MooseDocs algorithm extension

### ISSUE THREAD

**Issue by zachmprince** [idaholab/moose#20202]
## Reason

Since we generally all some sort of applied mathematician, it would probably behoove us to be able to cleanly write algorithms as part of our documentation.

## Design

I'm thinking of having something similar to the [latex algorithmicx package](https://ctan.math.illinois.edu/macros/latex/contrib/algorithmicx/algorithmicx.pdf) where the syntax could look something like this:

```
!algorithm
!function name=<name> param=<parameters>
!state <do stuff>
!function end
```

## Impact

Cool new documentation feature.

**Comment by YaqiWang** [comment]
For some reason, the PR did not close this automatically. Maybe the syntax in the PR's content is not accurate?

**Comment by zachmprince** [comment]
Closed via #20203
